pp/go/storage/head/services/persistener.go

```diff
@@ -0,0 +1,353 @@
+	"github.com/prometheus/prometheus/pp/go/util"
+)
+
+// defaultCoolingInterval the interval after which the rotation should have
+// taken place to eliminate errors in the selection from the catalog.
+const defaultCoolingInterval = 60 * time.Second
+
+//
+// Persistener
+//
+
+// Persistener converts and saves spent [Head]s.
+type Persistener[
+	TTask Task,
+	TShard, TGoShard Shard,
+	THeadBlockWriter HeadBlockWriter[TShard],
+	THead Head[TTask, TShard, TGoShard],
+] struct {
+	catalog       *catalog.Catalog
+	blockWriter   THeadBlockWriter
+	writeNotifier WriteNotifier
+
+	clock               clockwork.Clock
+	tsdbRetentionPeriod time.Duration
+	retentionPeriod     time.Duration
+	// stat
+	events                  *prometheus.CounterVec
+	headPersistenceDuration prometheus.Histogram
+}
+
+// NewPersistener init new [Persistener].
+func NewPersistener[
+	TTask Task,
+	TShard, TGoShard Shard,
+	THeadBlockWriter HeadBlockWriter[TShard],
+	THead Head[TTask, TShard, TGoShard],
+](
+	hcatalog *catalog.Catalog,
+	blockWriter THeadBlockWriter,
+	writeNotifier WriteNotifier,
+	clock clockwork.Clock,
+	tsdbRetentionPeriod time.Duration,
+	retentionPeriod time.Duration,
+	registerer prometheus.Registerer,
+) *Persistener[TTask, TShard, TGoShard, THeadBlockWriter, THead] {
+	factory := util.NewUnconflictRegisterer(registerer)
+	return &Persistener[TTask, TShard, TGoShard, THeadBlockWriter, THead]{
+		catalog:             hcatalog,
+		blockWriter:         blockWriter,
+		writeNotifier:       writeNotifier,
+		clock:               clock,
+		tsdbRetentionPeriod: tsdbRetentionPeriod,
+		retentionPeriod:     retentionPeriod,
+		events: factory.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "prompp_head_event_count",
+				Help: "Number of head events",
+			},
+			[]string{"type"},
+		),
+		headPersistenceDuration: factory.NewHistogram(
+			prometheus.HistogramOpts{
+				Name: "prompp_head_persistence_duration",
+				Help: "Block write duration in milliseconds.",
+				Buckets: []float64{
+					500, 1000, 2500, 5000, 7500,
+					10000, 25000, 50000, 75000, 100000,
+				},
+			},
+		),
+	}
+}
+
+// Persist spent [Head]s.
+//
+//revive:disable-next-line:function-length // long but readable.
+//revive:disable-next-line:cognitive-complexity // long but readable.
+//revive:disable-next-line:cyclomatic // long but readable.
+func (p *Persistener[TTask, TShard, TGoShard, THeadBlockWriter, THead]) Persist(heads []THead) (outdatedHeads []THead) {
+	shouldNotify := false
+	for _, head := range heads {
+		start := p.clock.Now()
+		if !head.IsReadOnly() {
+			continue
+		}
+
+		logger.Debugf("[Persistener]: head %s start persist", head.ID())
+		if record, err := p.catalog.Get(head.ID()); err != nil {
+			logger.Errorf("[Persistener]: failed get head %s from catalog: %v", head.ID(), err)
+		} else if record.Status() == catalog.StatusPersisted {
+			logger.Debugf("[Persistener]: persisted head %s is outdated", head.ID())
```

Comment:
This line should be in next condition. Here may be log like "head is already persisted. Skip it". But it is more likely further before continue command.